azext_edge/tests/edge/orchestration/test_work_unit.py

```diff
@@ -0,0 +1,775 @@
+
+
+path_pattern_base = r"^/subscriptions/[0-9a-fA-F-]+/resourcegroups/[a-zA-Z0-9]+"
+STANDARD_HEADERS = {"content-type": "application/json"}
+
+
+class CallKey(Enum):
+    CONNECT_RESOURCE_MANAGER = "connectResourceManager"
+    GET_CLUSTER = "getCluster"
+    GET_RESOURCE_PROVIDERS = "getResourceProviders"
+    DEPLOY_INIT_WHATIF = "deployInitWhatIf"
+    DEPLOY_INIT = "deployInit"
+    GET_SCHEMA_REGISTRY = "getSchemaRegistry"
+    GET_CLUSTER_EXTENSIONS = "getClusterExtensions"
+    GET_SCHEMA_REGISTRY_RA = "getSchemaRegistryRoleAssignments"
+    PUT_SCHEMA_REGISTRY_RA = "putSchemaRegistryRoleAssignment"
+    DEPLOY_CREATE_WHATIF = "deployCreateWhatIf"
+    DEPLOY_CREATE = "deployCreate"
+
+
+class RequestKPIs(NamedTuple):
+    method: str
+    url: str
+    params: dict
+    path_url: str
+    body_str: str
+
+
+class ExceptionMeta(NamedTuple):
+    exc_type: Type[Exception]
+    exc_msg: Union[str, List[str]] = ""
+
+
+class ServiceGenerator:
+    def __init__(
+        self,
+        scenario: dict,
+        mocked_responses: responses,
+    ):
+        self.scenario = scenario
+        self.mocked_responses = mocked_responses
+        self.call_map: Dict[CallKey, List[RequestKPIs]] = {}
+        self._bootstrap()
+
+    def _bootstrap(self):
+        omit_methods: Optional[FrozenSet[str]] = self.scenario.get("omitHttpMethods")
+        if not omit_methods:
+            omit_methods = frozenset([])
+        for method in [
+            responses.GET,
+            responses.HEAD,
+            responses.POST,
+            responses.PUT,
+        ]:
+            if method not in omit_methods:
+                self.mocked_responses.add_callback(
+                    method=method, url=re.compile(r".*"), callback=self._handle_requests
+                )
+        self._reset_call_map()
+
+    def _reset_call_map(self):
+        self.call_map = {}
+        for key in CallKey:
+            self.call_map[key] = []
+
+    def _handle_requests(self, request: requests.PreparedRequest) -> Optional[tuple]:
+        request_kpis = get_request_kpis(request)
+        for handler in [self._handle_common, self._handle_init, self._handle_create]:
+            handler_response = handler(request_kpis)
+            if handler_response:
+                return handler_response
+
+        raise RuntimeError(f"No match for {request_kpis.method} {request_kpis.url}.")
+
+    def _handle_common(self, request_kpis: RequestKPIs) -> Optional[tuple]:
+        # return (status_code, headers, body)
+        if request_kpis.method == responses.HEAD:
+            if request_kpis.url == ARM_ENDPOINT:
+                self.call_map[CallKey.CONNECT_RESOURCE_MANAGER].append(request_kpis)
+                return (200, {}, None)
+
+        if request_kpis.method == responses.GET:
+            if request_kpis.path_url == f"/subscriptions/{ZEROED_SUBSCRIPTION}/providers":
+                assert request_kpis.params["api-version"] == "2024-03-01"
```

Comment:
nit - thoughts on moving some of these API-versions to at least a local constant value? Benefit would be easier updates when api-versions change, but cost is that we might lose failed-test specificity if we have them all mapped to the same const

Reply:
Good thought, went for a local `ExpectedAPIVersion` enum.